src/app/services/courses-store.service.ts

```diff
@@ -1,44 +1,162 @@
-import { Injectable } from "@angular/core";
+import { Injectable } from '@angular/core';
+import { Course, FilterCourse } from '@app/shared/types/course.model';
+import { BehaviorSubject, Observable, tap } from 'rxjs';
+import { CoursesService } from './courses.service';
+import {
+  AllAuthorsResponse,
+  AllCourseResponse,
+  AuthorResponse,
+  CourseResponse,
+  CreateAuthorResponse,
+  CreateCourseResponse,
+  DeleteCourseResponse,
+  FilterResponse,
+} from './api.models';
+import { Author } from '@app/shared/types/author.model';
 
 @Injectable({
-  providedIn: "root",
+  providedIn: 'root',
 })
 export class CoursesStoreService {
+  private courses$$ = new BehaviorSubject<Course[]>([]);
+  public courses$ = this.courses$$.asObservable();
+  private authors$$ = new BehaviorSubject<Author[]>([]);
+  public authors$ = this.authors$$.asObservable();
+  private isLoading$$ = new BehaviorSubject<boolean>(false);
+  public isLoading$ = this.isLoading$$.asObservable();
+
+  constructor(private courseService: CoursesService) {}
+
   getAll() {
-    // Add your code here
+    this.isLoading$$.next(true);
+    this.courseService.getAll().subscribe({
+      next: (response: AllCourseResponse) => {
+        this.courses$$.next(response.result);
+        this.isLoading$$.next(false);
+      },
+      error: err => {
+        console.error('Failed to get courses', err);
+        this.isLoading$$.next(false);
+      },
+    });
   }
 
-  createCourse(course: any) {
-    // replace 'any' with the required interface
-    // Add your code here
+  createCourse(course: Omit<Course, 'id'>) {
+    this.isLoading$$.next(true);
+    this.courseService.createCourse(course).subscribe({
+      next: (response: CreateCourseResponse) => {
+        this.courses$$.next([...this.courses$$.value, response.result]);
```

Comment:
Well, theoretically this is fine, in a scalable application this might not be the case. Suppose that multiple courses are added by multiple users simultaneously. You are adding the new one to the existing list, which will not include the new courses created by others. Try to get the full courses list after a new one is created. Or you might even want to do that but get the full list once navigating back to courses page

Reply:
Yes, right. Actually conceptually it was not quite ok as get coursesAll and get authorsAll requests were sent in app component which is not good for sure for a not authorised user. (i sent these requests at login) -> All authors and courses are requested at courses ( home component for authorised users ) ngOnInit -> This way everytime we navigate back to the courses page the courses will be asked and refreshed.